src/io/rsm/rsm_client.hpp

```diff
@@ -131,6 +153,117 @@ class RsmClient {
 
     return TimedOut{};
   }
+
+  /// AsyncRead methods
+  void SendAsyncReadRequest(ReadRequestT req) {
+    MG_ASSERT(!async_read_);
+
+    ReadRequest<ReadRequestT> read_req;
+    read_req.operation = req;
+
+    if (!async_read_before_) {
+      async_read_before_ = io_.Now();
+    }
+    current_read_request_ = req;
+    async_read_ = io_.template Request<ReadRequest<ReadRequestT>, ReadResponse<ReadResponseT>>(leader_, read_req);
+  }
+
+  std::optional<BasicResult<TimedOut, ReadResponseT>> PollAsyncReadRequest() {
+    MG_ASSERT(async_read_);
+
+    if (!async_read_->IsReady()) {
+      return std::nullopt;
+    }
+
+    return AwaitAsyncReadRequest();
+  }
+
+  std::optional<BasicResult<TimedOut, ReadResponseT>> AwaitAsyncReadRequest() {
+    ResponseResult<ReadResponse<ReadResponseT>> get_response_result = std::move(*async_read_).Wait();
+    async_read_.reset();
+
+    const Duration overall_timeout = io_.GetDefaultTimeout();
+    const bool past_time_out = io_.Now() < *async_read_before_ + overall_timeout;
+    const bool result_has_error = get_response_result.HasError();
+
+    if (result_has_error && past_time_out) {
```

Comment:
We are not hiding the error. The only possible error type we can get from the `BasicResult<...>` is `TimeOut` error.
The `result_has_error` flag is indicating the `TimeOut` error we get from the future(if there is one).
The `past_time_out` flag is indicating if the time frame defined by the `RsmClient` is over or not.
That can mean that we are dealing with a redirection. And we can still retry at another address if we are within the time frame, defined by the `RsmClient`.
In short, if this condition is evaluating to true, then we propagate the presence of a `TimeOut` to the application code, otherwise retry at another address if we still have time to do so.

Reply:
>The only possible error type we can get from the BasicResult<...> is TimeOut error. 
Yep and I think that's an issue in the long run, because a TimeOut is not the only error we can get... Anyway, leave this for now, it was the same as before.